### ISSUE TITLE
Potentially incorrect type error for operator + for 2 lists

### ISSUE THREAD

**Issue by 108anup** [microsoft/pyright#3669]
**Code**
```
l1 = [1, 2, 3]
l2 = [[1, 2]]
l1 + l2
# output: [1, 2, 3, [1, 2]]
```

**Observed behavior**
Get type error: `Operator "+" not supported for types "list[int]" and "list[list[int]]"`

**Expected behavior**
Since this operation is allowed in python, perhaps there shouldn't be any type error.
The output list has type: `list[Union[list[int], int]]`.

**Comment by erictraut** [comment]
This is correct behavior, not a bug.

You have not declared a type for `l1` or `l2`, so their types are inferred. The inferred type of `l1` is `list[int]`, and the inferred type of `l2` is `list[list[int]]`. The `__add__` method for the `list` class is declared in a manner that does not allow you to mix different types of list. To do so would violate the type.

If you want to support the `+` operator for disparate types of lists, you would need to declare the type of `l1` differently. Here are a couple of options:

```python
l1: list[int | list[int]] = [1, 2, 3]
```

or 

```python
l1: list[Any] = [1, 2, 3]
```

For more details about type declarations and type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by 108anup** [comment]
Got it thanks.

Out of curiosity, would it ever make sense to infer type for `l1` as `list[int | list[int]]` based on the program sequence/context? I guess inferring such looser types might nullify the benefits of static type checking for inferred types.

**Comment by erictraut** [comment]
What you're suggesting might be feasible in languages that are less extensible than Python. But in Python, any type can override any operator. For example, the `+` operator in your example is implemented as a method `__add__` or `__radd__` or both. When evaluating an expression like `x + y`, a type checker must know the type of `x` and `y` beforehand to determine which method to apply and therefore whether the operation is supported. It's not feasible to do this in reverse (assume that `x + y` is valid and on that basis infer their types).
